### PR TITLE
Add PlayerDeathEvent->setKeepXp() as per #4002

### PR DESCRIPTION
**Pull request by TheNewHEROBRINEX**
## Introduction
This PR aims to reduce the complexity needed for a plugin to prevent loss of experience upon player death by adding the method setKeepXp to PlayerDeathEvent.

### Relevant issues

The need for this addition was expressed in #4002.

### API changes

Added PlayerDeathEvent->setKeepXp().

### Behavioural changes

None when no plugin is using the new API method.

## Backwards compatibility

Yes

## Follow-up

None

## Tests

This PR was tested using [this script plugin](https://gist.github.com/TheNewHEROBRINEX/fc01eceb0a6cabb068814343e344d14b) with the related video that can be seen [here](https://youtu.be/o2tqXO8zsfo).

### CODE REVIEW

**Comment by TheNewHEROBRINEX** [comment]
In the video, I also take damage on purpose before doing /kill to check that other attributes are still being reset properly.

File: src/event/player/PlayerDeathEvent.php [hunk 1]
```diff
@@ -41,6 +41,8 @@ class PlayerDeathEvent extends EntityDeathEvent{
 	private $deathMessage;
 	/** @var bool */
 	private $keepInventory = false;
+	/** @var bool */
+	private $keepXp = false;
```
Comment (dktapps) [inline src/event/player/PlayerDeathEvent.php:45]: You can use a native type for this. master is 7.4+ only now.

Comment (TheNewHEROBRINEX) [reply src/event/player/PlayerDeathEvent.php:45]: Do I have to keep the getter and the setter anyway for code consistency?

Comment (dktapps) [reply src/event/player/PlayerDeathEvent.php:45]: I'm not telling you to remove those.

Comment (TheNewHEROBRINEX) [reply src/event/player/PlayerDeathEvent.php:45]: I know, I was just asking

File: src/player/Player.php [hunk 2]
```diff
@@ -2135,6 +2138,10 @@ public function respawn() : void{
 		$this->setHealth($this->getMaxHealth());
 
 		foreach($this->attributeMap->getAll() as $attr){
+			if($attr->getId() === Attribute::EXPERIENCE or $attr->getId() === Attribute::EXPERIENCE_LEVEL){ //we have already reset both of those if needed when the player died
+				$this->getNetworkSession()->syncAttributes($this, [$attr]);
```
Comment (dktapps) [inline src/player/Player.php:2142]: why is this necessary?

Comment (TheNewHEROBRINEX) [reply src/player/Player.php:2142]: Because the client on respawn assumes vanilla behaviour and automatically resets experience progress and level to 0. Without this when your respawn it will appear that you have lost your experience but if you leave and join again it appears back.

Comment (dktapps) [reply src/player/Player.php:2142]: I think markSynchronized(false) should be sufficient.

Comment (dktapps) [reply src/player/Player.php:2142]: Note to self: this reset might also be causing some other problems (for example setting health to max before resetting attribute defaults might lead to health overflow).

**Comment by TheNewHEROBRINEX** [comment]
I tested the PR again and confirmed that it is working

**Review by dktapps — APPROVED** [review]
Tested and found working, apologies for the long delay ...
